### PR TITLE
Fix NavigationAgent3D not emitting "target_reached" Signal reliably

### LINKED ISSUE #47334: NavigationAgent3D signal "target_reached" fires very unreliable with continues movement
**Godot version:**
Godot v4.0.dev.custom_build
commit 4a1f2dcb746af43dcebeb241276adfb5d99c1020

**OS/device including version:**
Win 10 64x

**Issue description:**
NavigationAgent3D signal "target_reached" fires very unreliable with continues movement breaking behaviour that depends on this signal.

If the NavigationAgent3D doesn't stop and instead immediately proceeds with movement and follows a new path as soon as the path end is reached and "navigation_finished" signal is fired it constantly fails to fire the "target_reached" signal as well.

With very slow movement speed (something unpractical for most games like 2-4) and a way to large margin for realistic scaled characters (1.0+) it fails 8 out of 10 attempts to fire "target_reached" with human sized characters. It even fails if the agent moves directly over the target point while "navigation_finished" fires every single time.

**Steps to reproduce:**
Move a NavigationAgent3D constantly between multiple points. Change to an new navpath immediatly when path is finished / "navigation_finished" is fired. See that "target_reached" signal barely gets fired in this case even with generous margin values.

I think the issue is that the "target_reached" signal function only checks once on the slow physics_tick rate while "navigation_finished" signal function checks on every single path and next_location query from scripts.

This is actually easy to test by moving a script from _physics_process() to _process(). With _process(), high framerate and high movement speed "target_reached" stops firing completely while "navigation_finished" runs fine. This makes the issue worse with scripts that run in _process() to e.g. update a movement location or path marker UI each frame to prevent visual lag.

I suggest to add another "target_reached" signal check inside the "navigation_finished" check so both signals can fire reliably at the same time if the distance to the target is small enough.

**Minimal reproduction project:**
[NavigationAgent3DIssue.zip](https://github.com/godotengine/godot/files/6293312/NavigationAgent3DIssue.zip)

### PR DESCRIPTION
**Pull request by smix8**
Fixes #47334

NavigationAgent3D did only check distance to target on the slow physics tickrate. This created situations where a fast moving agent on high framerate would not emit the "target_reached" signal when it would immediatly move to the next target on completing the current navigation path (even if standing right on top the target position).

Now whenever the NavigationAgent3D completes a path follow and the  "navigation_finished" signal is emitted a target distance checks is performed to also emit "target_reached" if distance to target is small enough.

### CODE REVIEW

**Comment by smix8** [comment]
I don't use 2D or have a test project ready but I see now that NavigationAgent2D has the same code base on this issue.

I assume 2D navigation users will face the same issue on high framerates so I will prepare a pr for the NavigationAgent2D that applies the same fix.

**Comment by smix8** [comment]
@akien-mga 
Can this be merged? I am working on a PR to fix #47337 and other NavigationAgent3D bugs and documentation issues, and it would be cleaner for the signal and behavior documentation if this smaller fix could be part of master.

**Comment by smix8** [comment]
@AndreaCatania

File: scene/3d/navigation_agent_3d.cpp [hunk 1]
```diff
@@ -309,6 +309,12 @@ void NavigationAgent3D::update_navigation() {
 		while (o.distance_to(navigation_path[nav_path_index] - Vector3(0, navigation_height_offset, 0)) < target_desired_distance) {
 			nav_path_index += 1;
 			if (nav_path_index == navigation_path.size()) {
+				if (!target_reached) {
+					if (distance_to_target() < target_desired_distance) {
+						emit_signal("target_reached");
+						target_reached = true;
+					}
+				}
```
Comment (AndreaCatania) [inline scene/3d/navigation_agent_3d.cpp:317]: Can you please put this inside a function and call it here and here: https://github.com/godotengine/godot/blob/c10007d32164b17102dfc669bbbf14f87630ed34/scene/3d/navigation_agent_3d.cpp#L109

**Review by AndreaCatania — CHANGES_REQUESTED** [review]
Just a small change, but the code looks good. Please let me know if you can do the same for Agent 2D, it's not required for this PR though.

File: scene/3d/navigation_agent_3d.cpp (hunk 1 above)
Comment (smix8) [reply scene/3d/navigation_agent_3d.cpp:317]: Created a private `_check_distance_to_target` function for this.

**Comment by smix8** [comment]
Damn, sorry! Hit the wrong button on the wrong branch. No idea how to turn it back :(

**Comment by smix8** [comment]
Created a clean branch to remove this mess #48229

**Pull request closed without merging**